executor/index_lookup_hash_join.go

```diff
@@ -94,18 +94,19 @@ type indexHashJoinResult struct {
 
 type indexHashJoinTask struct {
 	*lookUpJoinTask
-	outerRowStatus []outerRowStatusFlag
+	outerRowStatus [][]outerRowStatusFlag
 	lookupMap      *rowHashMap
 	err            error
 	keepOuterOrder bool
 	// resultCh is only used when the outer order needs to be promised.
 	resultCh chan *indexHashJoinResult
 	// matchedInnerRowPtrs is only valid when the outer order needs to be
 	// promised. Otherwise, it will be nil.
-	// len(matchedInnerRowPtrs) equals to len(lookUpJoinTask.outerResult), and
-	// the elements of every row indicates the matched inner row ptrs of the
-	// corresponding outer row.
-	matchedInnerRowPtrs [][]chunk.RowPtr
+	// len(matchedInnerRowPtrs) equals to
+	// len(lookUpJoinTask.outerResult.NumChunks()), and the elements of every
+	// matchedInnerRowPtrs[chkIdx][rowIdx] indicates the matched inner row ptrs
+	// of the corresponding outer row.
+	matchedInnerRowPtrs [][][]chunk.RowPtr
```

Comment:
How about using the `map[chunk.RowPtr][]chunk.RowPtr` instead of the three-dimension array?

Reply:
We can not. `matchedInnerRowPtr` needs to keep the order, but map does not promise the order.